packages/neon/neon_files/lib/widgets/browser_view.dart

```diff
@@ -96,39 +83,32 @@ class _FilesBrowserViewState extends State<FilesBrowserView> {
-                                          stream: matchingDownloadTasks.isNotEmpty
-                                              ? matchingDownloadTasks.first.progress
-                                              : Stream.value(null),
-                                          builder: (final context, final downloadTaskProgressSnapshot) => _buildFile(
-                                            context: context,
-                                            details: FileDetails(
-                                              path: [...widget.bloc.path.value, file.name],
-                                              isDirectory: matchingUploadTasks.isEmpty && file.isDirectory,
-                                              size: matchingUploadTasks.isNotEmpty
-                                                  ? matchingUploadTasks.first.size
-                                                  : file.size,
-                                              etag: matchingUploadTasks.isNotEmpty ? null : file.etag,
-                                              mimeType: matchingUploadTasks.isNotEmpty ? null : file.mimeType,
-                                              lastModified: matchingUploadTasks.isNotEmpty
-                                                  ? matchingUploadTasks.first.lastModified
-                                                  : file.lastModified,
-                                              hasPreview: matchingUploadTasks.isNotEmpty ? null : file.hasPreview,
-                                              isFavorite: matchingUploadTasks.isNotEmpty ? null : file.favorite,
-                                            ),
-                                            uploadProgress: uploadTaskProgressSnapshot.data,
-                                            downloadProgress: downloadTaskProgressSnapshot.data,
-                                          ),
-                                        ),
+                                      final FileDetails details;
+                                      if (matchingDownloadTasks != null) {
+                                        details = FileDetails.downloading(
+                                          task: matchingDownloadTasks,
+                                          file: file,
+                                        );
+                                      } else if (matchingUploadTasks != null) {
+                                        details = FileDetails.uploading(
+                                          task: matchingUploadTasks,
+                                        );
+                                      } else {
+                                        details = FileDetails.fromWebDav(
+                                          file: file,
+                                          path: widget.bloc.path.value,
+                                        );
+                                      }
```

Comment:
I wonder if it would make sense to put this logic into a factory constructor :thinking: